File: LeetCode/1042. Flower Planting With No Adjacent/solution.py
"""
51 / 51 test cases passed.
Runtime: 108 ms
Memory Usage: 19.3 MB
"""
class Solution:
    def gardenNoAdj(self, n: int, paths: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(n)]
        ans = [0] * n
        for x, y in paths:
            graph[x - 1].append(y - 1)
            graph[y - 1].append(x - 1)
        for i in range(n):
            ans[i] = ({1, 2, 3, 4} - {ans[adj] for adj in graph[i]}).pop()
        return ans
        

File: LeetCode/1042. Flower Planting With No Adjacent/solution.cpp
/**
 * 51 / 51 test cases passed.
 * Runtime: 132 ms
 * Memory Usage: 48.4 MB 
 */
class Solution {
public:
    vector<int> gardenNoAdj(int n, vector<vector<int>>& paths) {
        unordered_map<int, vector<int>> graph;
        for (auto &path : paths) {
            graph[path[0] - 1].push_back(path[1] - 1);
            graph[path[1] - 1].push_back(path[0] - 1);
        }
        vector<int> ans(n);
        for (int i = 0; i < n; ++ i) {
            vector<bool> used(5);
            for (auto &adj: graph[i]) {
                used[ans[adj]] = true;
            }
            for (int j = 1; j <= 4; ++ j) {
                if (!used[j]) {
                    ans[i] = j;
                }
            }
        }
        return ans;
    }
};
